### ISSUE TITLE
[Feature] Handling of feature ID's and Json

### ISSUE THREAD

**Issue by maho7791** [osgi/osgi#266]
This discussion refers to Pull-Request: https://github.com/osgi/osgi/pull/261

For me the features ID, that is a GAV String, similar to what we do with `org.osgi.framework.Filter`. We have the ' FrameworkUtil` that is the factory for creating Filter instances, where Filter is an Interface. This factory provides the validation of the provided filter in respect to the syntax. So it takes care that the provided filter string is correct.
Beside this, we use the filter-strings in our service properties, that are later converted into Filter instances.

For the feature ID that would mean, making it an interface. We have to provide one or more factory methods, with corresponding signatures (maybe from the current class implementation), that validate the provided parameters and ensure that we get a valid ID instance. In this case it would be possible to delegate the implementation of the validation to specification implementors. The specification should then define the validation rules.

That would also make it possible to define the type of an ID in e.g. the `org.osgi.service.feature.FeatureArtifact` as a String to must be a valid ID-String. That would make the corresponding JSON, more compliant to common Json-Serializers. So a Json for a Feature artifact could always look like this:
```
{
  "id": "org.acme:acmeapp:1.0.1",
  "name": "The Acme Application",
  "license": "https://opensource.org/licenses/Apache-2.0",
  "isComplete": true,
  "bundles": [
        { "id": "org.osgi:org.osgi.util.function:1.1.0"},
        { "id": "org.osgi:org.osgi.util.promise:1.1.1"},
        {
            "id": "org.apache.commons:commons-email:1.5",
            "metadata":[
                 {"org.acme.javadoc.link": "https://commons.apache.org/proper/commons-email/javadocs/api-1.5"},
                 {"id": "test"}
             ]
      },
      { "id": "com.acme:acmelib:1.7.2"}      
    ]
}
```
instead of the format in the current. Because especially when using the metadata, this format would allow, having a property with key id in the metadata map. When having the valid Json like this:
```
{
  "id": "org.acme:acmeapp:1.0.1",
  "name": "The Acme Application",
  "license": "https://opensource.org/licenses/Apache-2.0",
  "isComplete": true,
  "bundles": [
      "id": "org.osgi:org.osgi.util.function:1.1.0",
      "id": "org.osgi:org.osgi.util.promise:1.1.1",
      {
          "id": "org.apache.commons:commons-email:1.5",
          "org.acme.javadoc.link": "https://commons.apache.org/proper/commons-email/javadocs/api-1.5",
          "id": "test"// HMMM, could lead into confusion
        },
        "id": "com.acme:acmelib:1.7.2"      
      ]
}
```
We may endup with additional restrictions and rules regarding serialization an de-serialization. If we handle it always as complex object, it would make the underlying model more clear. I think this is better than creating an own Json dialect , just for removing some boilerplate characters.

 I also respect, that we want to be close to the configurator format. In the case of features the structure is more complex an may get more complex in future versions.  Thats' why I think we shouldn't create a serialization format, that is causes additional efforts for customizing serializiers and de-serializers to fit our Json model. Such an additional complexity, just for reading or writing a Json-like files, may keep potential implementors away. That would have to spend a extra effort in custom de-/serialization instead of implementing functionality.

That's why I would discuss a more compliant Json format, that is easy to understand for users and implementors.

**Comment by bosschaert** [comment]
Bare IDs are no longer supported. They need to be inside a JSON object.

Custom metadata should be prefixed with reverse DNS names (as mentioned in the spec). Non-prefixed metadata is for OSGi-specified keys only. So there should be no clash with the `"id"` key.

**Comment by bjhargrave** [comment]
Fixed by #294
